### PR TITLE
Poisson-dev and support for BTD and Overlap Matrix in NEGF

### PR DESCRIPTION
**Pull request by AsymmetryChou**
try to merge poisson-dev to main

### CODE REVIEW

File: dptb/negf/device_property.py [hunk 1]
```diff
@@ -134,41 +136,74 @@ def cal_green_function(self, energy, kpoint, eta_device=0., block_tridiagonal=Tr
...
-            self.V = 0.
-        
-        if not hasattr(self, "hd") or not hasattr(self, "sd") or self.kpoint is None:
-            self.hd, self.sd, _, _, _, _ = self.hamiltonian.get_hs_device(kpoint, self.V, block_tridiagonal)
-            self.kpoint = torch.tensor(kpoint)
-        elif not torch.allclose(self.kpoint, torch.tensor(kpoint), atol=1e-5):
-            self.hd, self.sd, _, _, _, _ = self.hamiltonian.get_hs_device(kpoint, self.V, block_tridiagonal)
-            self.kpoint = torch.tensor(kpoint)
-
+            self.oldV = None
```
Comment (floatingCatty) [inline dptb/negf/device_property.py:153]: 这块判断old_V 的逻辑其实有点冗余，初始化的时候设置一个self.V，然后每次和输入的V做对比看看要不要更新就行了

File: dptb/negf/lead_property.py [hunk 2]
```diff
@@ -91,33 +111,86 @@ def self_energy(self, kpoint, energy, eta_lead: float=1e-5, method: str="Lopez-S
...
         if not isinstance(energy, torch.Tensor):
-            energy = torch.tensor(energy)
+            energy = torch.tensor(energy) # Energy relative to Ef
 
+        # if not hasattr(self, "HL"):
+        #TODO: check here whether it is necessary to calculate the self energy every time
+
+
+        if not self.useBloch:
+            if not hasattr(self, "HL") or abs(self.voltage_old-self.voltage)>1e-6 or max(abs(self.kpoint-torch.tensor(kpoint)))>1e-6:
```
Comment (floatingCatty) [inline dptb/negf/lead_property.py:123]: voltage_old 同理 old_V

File: dptb/nn/hr2hk.py [hunk 3]
```diff
@@ -45,6 +45,8 @@ def __init__(
         self.node_field = node_field
         self.out_field = out_field
 
+        self.atom_norbs = []
```
Comment (floatingCatty) [inline dptb/nn/hr2hk.py:48]: 这个OrbitalMapper里有的

File: dptb/negf/negf_hamiltonian_init.py [hunk 4]
```diff
@@ -92,82 +163,353 @@ def initialize(self, kpoints, block_tridiagnal=False):
...
+
+        if leftmost_size is None:
+            leftmost_atoms_index = np.where(structase.positions[:,2]==min(structase.positions[:,2]))[0]
+            leftmost_size = sum([self.h2k.atom_norbs[leftmost_atoms_index[i]] for i in range(len(leftmost_atoms_index))])
+        if rightmost_size is None:   
+            rightmost_atoms_index = np.where(structase.positions[:,2]==max(structase.positions[:,2]))[0]
+            rightmost_size = sum([self.h2k.atom_norbs[rightmost_atoms_index[i]] for i in range(len(rightmost_atoms_index))])
+        
+        subblocks = split_into_subblocks_optimized(HK[0],leftmost_size,rightmost_size)
+        if subblocks[0] < leftmost_size or subblocks[-1] < rightmost_size:
```
Comment (floatingCatty) [inline dptb/negf/negf_hamiltonian_init.py:473]: 这个fail的次数多不，我记得optimized总把两边优化没

File: dptb/negf/negf_hamiltonian_init.py [hunk 5]
```diff
@@ -92,82 +163,353 @@ def initialize(self, kpoints, block_tridiagnal=False):
...
+            subblocks = split_into_subblocks(HK[0],leftmost_size,rightmost_size)
+            log.info(msg="The optimized block tridiagonalization is not successful, \
+                     the original block tridiagonalization is used.")
+        subblocks = [0]+subblocks
+        
+        
+        for ik in range(HK.shape[0]):
```
Comment (floatingCatty) [inline dptb/negf/negf_hamiltonian_init.py:480]: 后面的时候或许需要把k axis cat到一块，方便RGF直接vmap GPU并行，现在可以不管

File: dptb/negf/negf_hamiltonian_init.py [hunk 6]
```diff
@@ -188,23 +530,38 @@ def get_hs_device(self, kpoint, V, block_tridiagonal=False):
...
+            hd_k, sd_k, hl_k, su_k, sl_k, hu_k = f["hd"][ik], f["sd"][ik], f["hl"][ik], f["su"][ik], f["sl"][ik], f["hu"][ik]
+
+            if V.shape == torch.Size([]):
+                allorb = sum([hd_k[i].shape[0] for i in range(len(hd_k))])
+                V = V.repeat(allorb)
+            V = torch.diag(V).cdouble()
+            counted = 0
+            for i in range(len(hd_k)): # TODO: this part may have probelms when V!=0
+                l_slice = slice(counted, counted+hd_k[i].shape[0])
+                hd_k[i] = hd_k[i] - V[l_slice,l_slice]@sd_k[i]
```
Comment (floatingCatty) [inline dptb/negf/negf_hamiltonian_init.py:554]: V会是一个dense的matrix嘛？

File: dptb/negf/negf_hamiltonian_init.py [hunk 7]
```diff
@@ -251,8 +624,262 @@ def device_norbs(self):
         """ 
         return the number of atoms in the device Hamiltonian
         """
-        return self.atom_norbs[self.device_id[0]:self.device_id[1]]
+        return self.h2k.atom_norbs[self.device_id[0]:self.device_id[1]]
```
Comment (floatingCatty) [inline dptb/negf/negf_hamiltonian_init.py:627]: 同上，可以用OrbitalMapper的替换

File: dptb/negf/poisson_init.py [hunk 8]
```diff
@@ -0,0 +1,391 @@
...
+import logging
+#eps0 = 8.854187817e-12 # in the unit of F/m
+# As length in deeptb is in the unit of Angstrom, the unit of eps0 is F/Angstrom
+eps0 = eps0*1e-10 # in the unit of F/Angstrom
+
+log = logging.getLogger(__name__)
+
+class Grid(object):
+    # define the grid in 3D space
+    def __init__(self,xg,yg,zg,xa,ya,za):
```
Comment (floatingCatty) [inline dptb/negf/poisson_init.py:18]: Grid都是方形的嘛？

File: dptb/negf/recursive_green_cal.py [hunk 9]
```diff
@@ -85,25 +85,30 @@ def recursive_gf_cal(energy, mat_l_list, mat_d_list, mat_u_list, sd, su, sl, s_i
     gru = [None for _ in range(num_of_matrices-1)]
     grd = [i.clone() for i in gr_left]  # Our glorious benefactor.
     g_trans = gr_left[len(gr_left) - 1].clone()
+    gr_lc = [gr_left[len(gr_left) - 1].clone()]
     for q in range(num_of_matrices - 2, -1, -1):  # Recursive algorithm
         grl[q] = grd[q + 1] @ mat_l_list[q] @ gr_left[q]  # (B5) We get the off-diagonal blocks for free.
         gru[q] = gr_left[q] @ mat_u_list[q] @ grd[q + 1]  # (B6) because we need .Tthem.T for the next calc:
         grd[q] = gr_left[q] + gr_left[q] @ mat_u_list[q] @ grl[q]  # (B4) I suppose I could also use the lower.
         g_trans = gr_left[q] @ mat_u_list[q] @ g_trans
-
```
Comment (floatingCatty) [inline dptb/negf/recursive_green_cal.py:93]: 这个是啥

File: dptb/negf/negf_hamiltonian_init.py [hunk 10]
```diff
@@ -48,19 +60,76 @@ class NEGFHamiltonianInit(object):
...
+    def __init__(self, 
+                 model: torch.nn.Module,
+                 AtomicData_options: dict, 
+                 structure: ase.Atoms,
+                 block_tridiagonal: bool,
+                 pbc_negf: List[bool],
+                 stru_options:dict, 
+                 unit: str,
+                 results_path:Optional[str]=None,
+                 torch_device: Union[str, torch.device]=torch.device('cpu')
```
[the remark this answers is not in the document]
Comment (AsymmetryChou) [reply dptb/negf/negf_hamiltonian_init.py:72]: 那这个字段会和NEGF器件区域的device相互冲突的。这里是有意改为torch_device。

File: dptb/negf/negf_hamiltonian_init.py (hunk 6 above)
Comment (AsymmetryChou) [reply dptb/negf/negf_hamiltonian_init.py:554]: 不是的，V是一个对角矩阵

File: dptb/negf/poisson_init.py (hunk 8 above)
Comment (AsymmetryChou) [reply dptb/negf/poisson_init.py:18]: 对的，目前只支持方形
